robot/impl/robot_reconfigure_test.go

```diff
@@ -2419,8 +2425,12 @@ func TestUpdateWeakDependents(t *testing.T) {
 
 	_, err = robot.ResourceByName(weak1Name)
 	test.That(t, err, test.ShouldNotBeNil)
+	// Assert that the explicit dependency was observed.
 	test.That(t, err.Error(), test.ShouldContainSubstring, "unresolved dependencies")
 
+	// Reconfigure without the explicit dependency. While also adding a second component that would
+	// have satisfied the dependency from the prior `weakCfg1`. Due to the weak dependency wildcard
+	// matcher, this `base1` component will be parsed as an implicit dependency to `weak1`.
```

Comment:
```suggestion
	// matcher, this `base1` component will be parsed as a weak dependency of `weak1`.
```
[nit] "implicit" is a slightly overloaded term here. We use "implicit dependency" to refer to a dependency returned by a resource's `Validate` method.

Reply:
Ah interesting, thanks for that context. FWIW, I dislike `weak` because that has a meaning in "resource graphs" aka memory management. And in Java/C++, the meaning of weak doesn't map to what weak means here.

So to re-state the definitions:
* A weak dependency is a dependency identified by [applying a matcher](https://github.com/viamrobotics/rdk/blob/b39c81bab8768a06be9780d577fdc0e27be53ed5/resource/resource_registry.go#L134) defined at [Resource registration](https://github.com/viamrobotics/rdk/blob/b39c81bab8768a06be9780d577fdc0e27be53ed5/robot/impl/robot_reconfigure_test.go#L2390).
* An implicit dependency is one returned as a string from [ConfigValidator/"ConvertedAttributes".Validate()](https://github.com/viamrobotics/rdk/blob/b39c81bab8768a06be9780d577fdc0e27be53ed5/resource/config.go#L286-L290).

Implicit dependencies seem to typically be strings from a (user) config ([for example](https://github.com/viamrobotics/rdk/blob/b39c81bab8768a06be9780d577fdc0e27be53ed5/services/baseremotecontrol/builtin/builtin.go#L59)). And weak dependencies are identified by a set of [matchers defined at resource registration](https://github.com/viamrobotics/rdk/blob/b39c81bab8768a06be9780d577fdc0e27be53ed5/services/motion/builtin/builtin.go#L40-L43).

Can you confirm the following:
* Implicit dependencies and weak dependencies are ultimately both just individual resource nodes on the graph? Whether it be a component or service or otherwise.
* Are there any behavior differences in the viam server based on how dependencies work based on whether it was a weak vs implicit dependency? Or explicit for that matter? Or is a dependency in the resource graph just a dependency regardless of how it was identified.

Reply:
> FWIW, I disliked weak because that has a meaning in "resource graphs" aka memory management. And in Java/C++, the meaning of weak doesn't map to what weak means here. So I was intentionally using the word that felt more natural.

Ah good to know; I've never heard that term (and haven't written much Java/C++) and will look into it.

Those definitions you stated are exactly correct from what I understand.

> Implicit dependencies and weak dependencies are ultimately both just individual resource nodes on the graph? Whether it be a component or service or otherwise.

A dependency is really a directed edge in the resource graph that connects any two resource nodes. Those nodes could be components, services or remotes. Modules, processes, and packages are stored in their own managers.

> Are there any behavior differences in the viam server based on how dependencies work based on whether it was a weak vs implicit dependency? Or explicit for that matter? Or is a dependency in the resource graph just a dependency regardless of how it was identified.

Not that I know of. Weak, implicit and explicit seem to only differ in how they are set up (resource matcher, `Validate`, and `depends_on`) and not in their actual manifestation in the resource graph.

Reply:
Awesome, thanks Benji!

> Ah good to know; I've never heard that term (and haven't written much Java/C++) and will look into it.
In java, I assume you're familiar that memory management is automatic and everything is garbage collected. So when the last reference to an object disappears, that object can be cleaned up by a later garbage collection round. In [java](https://docs.oracle.com/javase/8/docs/api/java/lang/ref/WeakReference.html) a `WeakReference` is a way to keep a pointer/handle on an object without preventing it from being garbage collected. C++ with smart pointers refers to the same [idea as a](https://en.cppreference.com/w/cpp/memory/weak_ptr) `weak_ptr`.

Reply:
Ah gotcha; I've only heard of `shared_ptr`s  in C++, but that concept makes sense.